### PR TITLE
components: Add New Workspace Rollout Job

### LINKED ISSUE #15068: Epic: Automate blue/green deployments of workspace clusters
Currently workspace cluster blue/green deployments, including preparation, validation and traffic-shifting, are manual which results in a lot of person-hours wasted. As we are growing the number of Gitpod installations we manage, we need to reduce that burden. 

We are aiming to have in-place workspace application roll-outs (#15065). Although we know we will still have the need for blue/green deployments at least for infrastructure related changes. Therefore, we still need to invest in their automation.

## Value

1. Enable Gitpod Dedicated growth
2. Reduce engineering effort wasted on manual deployment and traffic shift

## Success Criteria

1. Full workspace roll-outs of multiple Gitpod installations, where entirely new workspace clusters are deployed, are achievable with a click of button
3. Roll-out success is automatically monitored, and in case of issues roll-back is automatically initiated
4. Manual rollback should be similarly possible
3. Roll-outs and roll-backs result in no service disruption
4. What we build is applicable to both Gitpod Dedicated (AWS) and gitpod.io (GCP). We need to be mindful of how we roll out dedicated and ensure it is compatible with what is built here. 

## Dependencies

N/A

## Plan

TBD

## Resources

[Upgrading Gitpod in a Dedicated world - Gap analysis (internal)](https://www.notion.so/gitpod/Upgrading-Gitpod-in-a-Dedicated-world-Gap-analysis-dbb143176e3c4484b5c617e73be4206c)

## Open questions

1. What are the most time consuming parts of the deployment?
3. Given (1), how can we decompose this epic into smaller projects?

### PR DESCRIPTION
**Pull request by Pothulapati**
## Description

This PR adds a new component called the `workspace-rollout-job`, That can automatically perform
a rollout on a meta cluster that has 2 workspace clusters registered. This is done while monitoring
the workspace success rate on the new cluster. A revert of the rollout is performed incase the metric
analysis returns that there is something wrong.

This job can be ran both inside and outside of a cluster, where we port-forward both the
`ws-manager-bridge` to perform the action. `prometheus-url` argument should
be used to give a prometheus URL for querying

The component is built by the following packages:

### Metric Analysis

This package includes a `Analyzer` Interface that has a `MoveForward` function
which can return one of the following values:

`-1`: Denoting that there is something wrong in the new cluster.
`1`  : Denoting that the rollout is going on, positively.
`0` : No Data is present to make a decision.

The current main implementation of this interface is the `WorkspaceKeyMetricsAnalyzer`,
which depends on the failure and total rate of the following key metrics, to calculate
the success ratio and compare that with a user provided target success ratio too
decide whether to move forward or not:

- Workspace Starts Total & Failures
- Workspace Stops Total & Failure
- Backups Total & Failure

This follows the same SLO's model that we already have in place and use, so that
we are on the same page with what's happening but focusing only on the
`workspaces` health and not image builds or prebuilds.

### Rollout

Rollout is the core component which performs the rollout. It spins up the above Analysis
routine in a separate go routine and runs it every `AnalysisWaitDuration`, and tries to perform
the rollout in steps with a wait of `rolloutWaitDuration`. The analysis and rollout are separated
so that we can have a more frequent analysis of Health (and there by revert in case of failure)
than the rollout step wait period.

If Analysis finds something wrong `r.Revert` channel is signalled, after which a revert of the rollout
is performed in a single action.

This rollout also exposes some metrics to pass on the state of the rollout process.

### `wsbridge`

`ws-manager-bridge` is a helper package that helps with the communication and
performing the score update action.

## Related Issue(s)

Fixes #15068 

## How to test

Use https://github.com/gitpod-io/workspace-preview/pull/76, to create a dev environment
and follow https://www.loom.com/share/f042061bcade4baea2295dd134234cb1

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/workspace-rollout-job/pkg/analysis/analyzer.go
Comment (ArthurSens) [inline components/workspace-rollout-job/pkg/analysis/analyzer.go:16]: This can definitely be done in another PR (if you agree that it's a good idea), but what do you think about replacing those raw ints into a type?

Something like:

```go
type Decision int

const (
  DecisionRevert		Decision = -1
  DecisionMoveForward		Decision = 1
  DecisionInsuficientData	Decision = 0
)

type Analyzer interface {
	// Given a cluster name, MoveForward is called by the rollout routine
	// repeatedly to determine whether to move forward on the rollout or not.
	GetDecision(ctx context.Context, clusterName string) (Decision, error)
}
```

Everything would work exactly the same, the only difference is that the code will become more readable in `pkg/rollout/rollout.go`

```go
decision, err := r.analyzer.GetDecision(context.Background(), r.newCluster)

if decision == analyzer.DecisionRevert {
	r.revert <- true
	return
} else if decision == analyzer.DecisionInsuficientData && r.currentScore >= r.okayScoreUntilNoData  {
	r.revert <- true
	return
} 
log.Info("Analyzer says yes, rolling forward")

```
_(A bit of code was deleted just to make this comment more compact)_

File: components/workspace-rollout-job/pkg/rollout/metrics.go
Comment (ArthurSens) [inline components/workspace-rollout-job/pkg/rollout/metrics.go:9]: It's really awesome that you want to be able to monitor the rollout job!

As we discussed in slack, metrics might not be the best approach here. What do you think about deleting all code related to metrics and relying entirely on logs?

File: components/workspace-rollout-job/pkg/rollout/rollout.go
Comment (ArthurSens) [inline components/workspace-rollout-job/pkg/rollout/rollout.go:54]: I believe all components from Gitpod use our own common-go/log library for our logs. Would it make sense to use it as well just to make log format consistent with all other Gitpod components?

File: components/workspace-rollout-job/pkg/rollout/rollout.go
Comment (ArthurSens) [inline components/workspace-rollout-job/pkg/rollout/rollout.go:104]: We could make code a little bit more readable if we decrease the indentation level 🙂:

```go
// Analyzer says no, stop the rollout
if moveForward == -1 {
	log.Info("Analyzer says no, stopping the rollout")
	r.revert <- true
	return
} else if moveForward == 0 && if r.currentScore >= r.okayScoreUntilNoData {
	log.Info("Current score is more than okayScoreUntilNoData, reverting the rollout")
	r.revert <- true
	return
}
// Roll forward otherwise
log.Info("Analyzer says yes, rolling forward")

```

File: components/workspace-rollout-job/pkg/wsbridge/wsbridge.go
Comment (ArthurSens) [inline components/workspace-rollout-job/pkg/wsbridge/wsbridge.go:46]: When doing port-forward manually with kubectl, sometimes I see that it just stops working all of a sudden and only works again if I stop and port-forward again. Is that scenario possible here too?

What would happen to the job if the port-forward suddenly stops working?

**Review by ArthurSens — APPROVED** [review]
Really awesome to finally see the effort getting into shape!

I've made a few comments regarding code readability and some really small improvement suggestions. I couldn't see anything that should block the merge, so I'm already approving the PR 🚀 

I'm adding a hold just in case you still want to make changes here, but feel free to address the comments in another PR if that makes sense to you

/hold

**Review by WVerlaek — COMMENTED** [review]
I've only had a brief look through the first files and found a small nit, but nice work! Looking forward to having this rollout job.

File: components/workspace-rollout-job/README.md
Comment (Pothulapati) [reply components/workspace-rollout-job/README.md:8]: README was out of date from all the latest changes. Just updated.

File: components/workspace-rollout-job/cmd/root.go
Comment (Pothulapati) [reply components/workspace-rollout-job/cmd/root.go:71]: The purpose of the server is to expose prometheus metrics!

File: components/workspace-rollout-job/pkg/analysis/analyzer.go
Comment (Pothulapati) [reply components/workspace-rollout-job/pkg/analysis/analyzer.go:16]: Definitely much better! Doing this for sure!

File: components/workspace-rollout-job/pkg/wsbridge/wsbridge.go
Comment (Pothulapati) [reply components/workspace-rollout-job/pkg/wsbridge/wsbridge.go:46]: Good Question!

> Is that scenario possible here too?

From my experiments even both shorter and longer, I have not experienced this with the job.

> What would happen to the job if the port-forward suddenly stops working?

The `UpdateScore` function would return an error, and thus causing the job to crash with a failed status (at that point which the rollout DRI should take manual action) . Considering that this is a failure with `wsbridge` and not prometheus, We don't have much else to do as these errors could be anything and not just port-forward-ing. For these kind of errors with `prometheus`, We handle them much more gracefully by reverting the rollout. 🤔

File: components/workspace-rollout-job/pkg/analysis/analyzer.go
Comment (Pothulapati) [reply components/workspace-rollout-job/pkg/analysis/analyzer.go]: Updated!

File: components/workspace-rollout-job/pkg/rollout/rollout.go
Comment (Pothulapati) [reply components/workspace-rollout-job/pkg/rollout/rollout.go:54]: Updated! 👍🏼

File: components/workspace-rollout-job/cmd/root.go
Comment (Pothulapati) [reply components/workspace-rollout-job/cmd/root.go]: Updated!

**Review by WVerlaek — APPROVED** [review]
Have 2 comments/questions, but the PR LGTM otherwise!

Adding a hold if you want to handle the comments and/or for Arthur to take another look as well
/hold

**Review by ArthurSens — APPROVED** [review]
Looks good to me, let's try it out and iterate over the feedback we get in the next rollout

File: components/workspace-rollout-job/pkg/rollout/rollout.go
Comment (Pothulapati) [reply components/workspace-rollout-job/pkg/rollout/rollout.go]: Good Find! Also, Updated a unit test to cover this!

File: components/workspace-rollout-job/pkg/analysis/analyzer.go
Comment (ArthurSens) [inline components/workspace-rollout-job/pkg/analysis/analyzer.go:24]: We can probably remove those comments now that we're not using -1, 0 and 1. The types are self explanatory 😛

File: components/workspace-rollout-job/cmd/root.go
Comment (Pothulapati) [reply components/workspace-rollout-job/cmd/root.go:96]: Yep!

**Pull request merged**
